### PR TITLE
refactor!: Consolidate CLI and run scripts

### PR DESCRIPTION
**Pull request by IvayloAndonov**
# Context

In reviewing [feedback](https://github.com/input-output-hk/cardano-js-sdk/pull/348#discussion_r927731133) from @Yuriy regarding a single entrypoint that can handle both program args and environment variables, we've found a solution to what led us down the path of separate run scripts, so we c now consolidate into just the `Commander.js` entrypoint.

This approach will improve our entrypoint documentation, given it’s then contained within `--help`, and also opens up the ability to use conflicts to simplify our logic between service discovery and static. This is a larger refactor, which I estimate to be worth an additional 2 points.

# Proposed Solution
- run.ts and startWorker.ts removed, with all ENVs represented in the cli.ts
- Providing DNS configuration + static for each service depdendency throws an error
- All missing required arguments/envs are reported specifically to the user

**_Cardano Services CLI help display:_** 
Screenshot 2022-08-22 at 2 25 36

**_Updated README file with usage examples:_** 
Screenshot 2022-08-29 at 10 57 56

Screenshot 2022-08-29 at 10 57 39

Screenshot 2022-08-22 at 17 10 49

# Important Changes Introduced
- refactor SDK's entry points, consolidate `CLI` and `run` scripts
- use Commander's custom [options](https://github.com/tj/commander.js#more-configuration) to allow centralized interaction with both args and env variables
- fix `withDb` decorator to use the correct service name in case of error
- update `Dockerfile` entry points and build scripts
- remove obsolete `run.ts`, `startWorker.ts` and related `scripts`
- extend `entrypoints.test.ts` and `entrypoints.txWorker.test.ts` files based on the introduced entry point level validations
- updated [README](https://github.com/input-output-hk/cardano-js-sdk/pull/394/files#diff-8190db84f584e67953bdc740053321b6a4d4ede419843a7bd47f90bab2729fc4) file, added usage examples

### CODE REVIEW

**Comment by IvayloAndonov** [comment]
I rebased on the top of _master_ for a clean state, ready for reviews.

Please pay attention to [entrypoints.test.ts](https://github.com/input-output-hk/cardano-js-sdk/pull/394/files#diff-71fc008b2e0656025c44e255b88caeb719b87d2d24ce40cbab8349614bef883a) which is significantly extended to cover all new entry point level validations and use cases.

**Review by iccicci — DISMISSED** [review]
Great!

File: packages/cardano-services/test/entrypoints.test.ts [hunk 1]
```diff
@@ -120,289 +135,1187 @@ describe('entrypoints', () => {
...
+        });
+      });
+
+      describe('specifying a PostgreSQL-dependent service', () => {
+        let spy: jest.Mock;
+        beforeEach(() => {
+          spy = jest.fn();
+        });
+
+        describe('without provided static nor service discovery config', () => {
```
Comment (rhyslbw) [inline packages/cardano-services/test/entrypoints.test.ts:213]: Missing `asset` scenario

File: packages/cardano-services/test/entrypoints.test.ts [hunk 2]
```diff
@@ -120,289 +135,1187 @@ describe('entrypoints', () => {
...
-              apiUrl,
-              '--logger-min-severity',
-              'error',
-              '--cardano-node-config-path',
-              cardanoNodeConfigPath,
-              ServiceNames.Utxo
-            ],
-            {
+        describe('with provided service discovery config', () => {
+          it('throws DNS SRV error and exits with code 1 when using CLI options', (done) => {
```
Comment (rhyslbw) [inline packages/cardano-services/test/entrypoints.test.ts:421]:
```suggestion
          it('exits with code 1 if DNS server not reachable when using CLI options', (done) => {
```

File: packages/cardano-services/test/entrypoints.test.ts [hunk 3]
```diff
@@ -120,289 +135,1187 @@ describe('entrypoints', () => {
...
+            });
+
+            proc.on('exit', (code) => {
+              expect(code).toBe(1);
+              expect(spy).toHaveBeenCalled();
+              done();
+            });
+          });
+
+          it('throws DNS SRV error and exits with code 1 when using env variables', (done) => {
```
Comment (rhyslbw) [inline packages/cardano-services/test/entrypoints.test.ts:461]:
```suggestion
          it(exits with code 1 if DNS server not reachable when using env variables', (done) => {
```

**Review by rhyslbw — CHANGES_REQUESTED** [review]
Looks great @IvayloAndonov, although the CLI tests can be improved here by implementing a couple of functions to reduce the duplication and highlight the scenario, given the expectations often are identical (just expecting an exit). You don't have to cover all of them, just the ones that are almost identical, where it's just different command line args or envs being passed

**Comment by IvayloAndonov** [comment]
> Looks great @IvayloAndonov, although the CLI tests can be improved here by implementing a couple of functions to reduce the duplication and highlight the scenario, given the expectations often are identical (just expecting an exit). You don't have to cover all of them, just the ones that are almost identical, where it's just different command line args or envs being passed

- README feedback is applied under https://github.com/input-output-hk/cardano-js-sdk/pull/394/commits/0acbdb6696672e321dd5a516bcde31da5cd4d850
- tests feedback is applied under https://github.com/input-output-hk/cardano-js-sdk/pull/394/commits/47a2887e868d0e91d816ad266adb170829abc1c3

**Review by mkazlauskas — DISMISSED** [review]
Nice refactor @IvayloAndonov :tada:

File: packages/cardano-services/test/entrypoints.test.ts [hunk 4]
```diff
@@ -24,6 +29,13 @@ const assertServiceHealthy = async (apiUrl: string, serviceName: ServiceNames) =
   expect(res.data).toEqual({ ok: true });
 };
 
+const getCommandLineOptionsWithoutDb = (serviceName: ServiceNames) => [
```
Comment (rhyslbw) [inline packages/cardano-services/test/entrypoints.test.ts:32]: This is a step in the right direction, but it's not reducing duplication that much, and it leaves the tests a little awkward to read. Try this:

```ts
const callCliAndAssertExit = (
  { args, env = {}, expectedExitCode }: { args: string[]; env?: NodeJS.ProcessEnv; expectedExitCode: number },
  done: jest.DoneCallback
) => {
  const spy = jest.fn();
  expect.assertions(2);
  const proc = fork(exePath('cli'), ['start-server', '--logger-min-severity', 'error', ...args], {
    env,
    stdio: 'pipe'
  });
  proc.stderr!.on('data', spy);
  proc.on('exit', (code) => {
    expect(code).toBe(expectedExitCode);
    expect(spy).toHaveBeenCalled();
    done();
  });
};
... 
         it('stake-pool exits with code 1', (done) => {
            callCliAndAssertExit({
              args: ['--services', ServiceNames.StakePool],
              expectedExitCode: 1
            },
            done);
          });
```

**Review by rhyslbw — CHANGES_REQUESTED** [review]

File: packages/cardano-services/package.json [hunk 5]
```diff
@@ -34,7 +34,7 @@
   ],
   "license": "Apache-2.0",
   "scripts": {
-    "build:cjs": "tsc --build src && shx chmod +x ./dist/cjs/cli.js ./dist/cjs/run.js ./dist/cjs/startWorker.js",
+    "build:cjs": "tsc --build src && shx chmod +x ./dist/cjs/cli.js",
```
[the remark this answers is not in the document]
Comment (rhyslbw) [reply packages/cardano-services/package.json:37]: Right, no need to include it in scope here. When it bites we can fix, but it's low priority.

File: packages/cardano-services/test/entrypoints.test.ts (hunk 4 above)
Comment (IvayloAndonov) [reply packages/cardano-services/test/entrypoints.test.ts]: Tests optimization is applied under 39834e029599adf5a77143c3fdb759ba46e36976

File: packages/cardano-services/test/entrypoints.test.ts [hunk 6]
```diff
@@ -24,6 +30,38 @@ const assertServiceHealthy = async (apiUrl: string, serviceName: ServiceNames) =
...
+  args?: string[];
+  env?: NodeJS.ProcessEnv;
+};
+
+const callCliAndAssertExit = (
+  { expectedExitCode, args = [], env = {}, dataMatchOnError, assertionsCount }: CallCliAndAssertExitArgs,
+  done: jest.DoneCallback
+) => {
+  const spy = jest.fn();
+  expect.assertions(assertionsCount);
```
Comment (rhyslbw) [inline packages/cardano-services/test/entrypoints.test.ts:46]:
```suggestion
  expect.assertions(dataMatchOnError ? 3 : 2 );
```
The expected assertions are deterministic right? Best to keep this an internal concern

Comment (rhyslbw) [reply packages/cardano-services/test/entrypoints.test.ts]: Done in b2ff0d7cc6e56dda7c265fdb4195c0e1080975e4

File: packages/cardano-services/README.md [hunk 7]
```diff
@@ -44,14 +41,25 @@ _The following examples require the [install and build] steps to be completed._
...
 ./dist/cjs/cli.js \
   start-server \
     --api-url http://localhost:6000 \
-    --cardano-node-config-path ./config/cardano-node/config.json \
-    --ogmios-url  ws://localhost:1338 \
+    --cardano-node-config-path ./config/network/testnet/cardano-node/config.json \
     --postgres-connection-string postgresql://somePgUser:somePassword@localhost:5432/someDbName \
-    asset,chain-history,stake-pool,tx-submit,network-info,utxo
+    --ogmios-url  ws://localhost:1338 \
+    asset,chain-history,stake-pool,tx-submit,network-info,utxo,rewards
```
Comment (iadmytro) [inline packages/cardano-services/README.md:52]: comma separated service names here parsed as one long service name
```suggestion
    asset chain-history stake-pool tx-submit network-info utxo rewards
```

File: packages/cardano-services/README.md [hunk 8]
```diff
@@ -64,28 +72,54 @@ _The following examples require the [install and build] steps to be completed._
...
+    --cardano-node-config-path ./config/network/testnet/cardano-node/config.json \
     --postgres-srv-service-name \
+    --postgres-db someDbName \
     --postgres-user somePgUser \
+    --postgres-password somePassword \
+    --ogmios-srv-service-name  some-domain-for-ogmios \
     --rabbitmq-srv-service-name some-domain-for-rabbitmq \
     --use-queue \
-    asset,chain-history,stake-pool,tx-submit,network-info,utxo
+    asset,chain-history,stake-pool,tx-submit,network-info,utxo,rewards
```
Comment (iadmytro) [inline packages/cardano-services/README.md:89]:
```suggestion
    asset chain-history stake-pool tx-submit network-info utxo rewards
```

File: packages/cardano-services/test/entrypoints.test.ts [hunk 9]
```diff
@@ -114,17 +162,15 @@ describe('entrypoints', () => {
           await ogmiosServerReady(ogmiosConnection);
         });
 
-        it('cli:start-server exposes a HTTP server at the configured URL with all services attached', async () => {
+        it('exposes a HTTP server at the configured URL with all services attached when using CLI options', async () => {
```
Comment (iadmytro) [inline packages/cardano-services/test/entrypoints.test.ts:165]: missing case for `enable-metrics`

**Review by iadmytro — CHANGES_REQUESTED** [review]

File: packages/cardano-services/README.md (hunk 8 above)
Comment (IvayloAndonov) [reply packages/cardano-services/README.md]: That wrong CLI usage in readme was introduced before my change so I assume it was due to misleading similarity between passing service names through `cli` and `run` .ts (run.ts uses commas or space separation).

Done in 8a21af9a2376eb03073ef97c909f5514dff135c4 @iadmytro

File: packages/cardano-services/test/entrypoints.test.ts (hunk 9 above)
Comment (IvayloAndonov) [reply packages/cardano-services/test/entrypoints.test.ts:175]: Done in 5172447ed17514d9dfb6b5e398b8f6a7e03044b5 @iadmytro

**Review by iadmytro — APPROVED** [review]

**Review by rhyslbw — APPROVED** [review]

**Review by mkazlauskas — APPROVED** [review]

**Pull request merged**